src/relay/ir/hash.cc

```diff
@@ -0,0 +1,335 @@
+    hash = Combine(hash, Hash(tensor_type->shape));
+    return hash;
+  }
+
+  size_t VisitType_(const IncompleteTypeNode* incomplete) final {
+    return GetRef<IncompleteType>(incomplete);
+  }
+
+  size_t VisitType_(const TypeVarNode* lhs) final {
+    return 0;
+    // if (const TypeVarNode* rhs = other.as<TypeVarNode>()) {
+    //   if (lhs->kind != rhs->kind) return false;
+    //   return LeafNodeEqual(GetRef<NodeRef>(lhs), other);
+    // } else {
+    //   return false;
+    // }
+  }
+
+  size_t VisitType_(const FuncTypeNode* func_type) final {
+    size_t hash = std::hash<std::string>()(func_type->_type_key);
+    for (auto type_param : func_type->type_params) {
+      hash = Combine(hash, TypeHash(type_param));
+    }
+
+    for (auto arg : func_type->arg_types) {
+      hash = Combine(hash, TypeHash(arg));
+    }
+
+    hash = Combine(hash, TypeHash(func_type->ret_type));
+    for (auto cs : func_type->type_constraints) {
+      hash = Combine(hash, TypeHash(cs));
+    }
+
+    return hash;
+  }
+
+  size_t VisitType_(const TypeRelationNode* type_rel) final {
+    return GetRef<TypeRelation>(type_rel).hash();
+    // if (const TypeRelationNode* rhs = other.as<TypeRelationNode>()) {
+    //   if (lhs->func->name != rhs->func->name) return false;
+    //   if (lhs->num_inputs != rhs->num_inputs) return false;
+    //   if (!this->AttrEqual(lhs->attrs, rhs->attrs)) return false;
+    //   if (lhs->args.size() != rhs->args.size()) return false;
+    //   for (size_t i = 0; i < lhs->args.size(); ++i) {
+    //     if (!TypeEqual(lhs->args[i], rhs->args[i])) return false;
+    //   }
+    //   return true;
+    // } else {
+    //   return false;
+    // }
+  }
+
+  size_t VisitType_(const TupleTypeNode* tuple_type) final {
+    size_t hash = std::hash<std::string>()(tuple_type->_type_key);
+    for (size_t i = 0; i < tuple_type->fields.size(); i++) {
+      hash = Combine(hash, TypeHash(tuple_type->fields[i]));
+    }
+    return hash;
+  }
+
+  // Expr equal checking.
+  size_t NDArrayHash(const runtime::NDArray& array) {
+    return 0;
+    // if (lhs.defined() != rhs.defined()) {
+    //   return false;
+    // } else if (lhs.same_as(rhs)) {
+    //   return true;
+    // } else {
+    //   auto ldt = lhs->dtype;
+    //   auto rdt = rhs->dtype;
+    //   CHECK_EQ(lhs->ctx.device_type, kDLCPU) << "can only compare CPU tensor";
+    //   CHECK_EQ(rhs->ctx.device_type, kDLCPU) << "can only compare CPU tensor";
+    //   if (ldt.code == rdt.code && ldt.lanes == rdt.lanes && ldt.bits == rdt.bits) {
+    //     size_t data_size = runtime::GetDataSize(*lhs.operator->());
+    //     return std::memcmp(lhs->data, rhs->data, data_size) == 0;
+    //   } else {
+    //     return false;
+    //   }
+    // }
+  }
+
+  int BindVar(const NodeRef& var) {
+    var_map_[var] = var_counter;
+    return var_counter++;
+  }
+
+  size_t VisitExpr_(const VarNode* var) final {
+    return std::hash<int>()(var_map_[GetRef<Var>(var)]);
```

Comment:
use .at()
there is the other case (hashing term with free var), it should be error with at() or just hash by ptr